server/src/server.ts

```diff
@@ -95,19 +97,22 @@ let findBinaryDirPathFromProjectRoot = (
     return null;
   }
 
-  return findBinaryDirPathFromProjectRoot(parentDir);
+  return findBinaryFromProjectDirRec(parentDir, targetDir, binaryName);
 };
 
-let getBinaryDirPath = (projectRootPath: p.DocumentUri) =>
-  extensionConfiguration.binaryPath == null
-    ? findBinaryDirPathFromProjectRoot(projectRootPath)
-    : extensionConfiguration.binaryPath;
-
 let findRescriptBinary = (projectRootPath: p.DocumentUri) =>
-  utils.findRescriptBinary(getBinaryDirPath(projectRootPath));
+  extensionConfiguration.binaryPath == null
+    ? findBinaryFromProjectDirRec(
+        projectRootPath,
+        c.nodeModulesBinDir,
```

Comment:
This, and the next parameter, look very specific.
That's the kind of thing I would see directly inside `findBinaryFromProjectDirRec`, not a parameter if only we could reduce the number of callers to `findBinaryFromProjectDirRec` to be only one.
See below.